src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunFileLayoutOptimizer.cs

```diff
@@ -0,0 +1,187 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+using Internal.TypeSystem;
+
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysis.ReadyToRun;
+using ILCompiler.DependencyAnalysisFramework;
+using System.Linq;
+using System.Collections.Immutable;
+using System.Text;
+using System.Reflection.Metadata.Ecma335;
+
+namespace ILCompiler
+{
+    public enum ReadyToRunMethodLayoutAlgorithm
+    {
+        DefaultSort,
+        ExclusiveWeight,
+        HotCold,
+        HotWarmCold
+    }
+
+    public enum ReadyToRunFileLayoutAlgorithm
+    {
+        DefaultSort,
+        MethodOrder
+    }
+
+    class ReadyToRunFileLayoutOptimizer
+    {
+        public ReadyToRunFileLayoutOptimizer (ReadyToRunMethodLayoutAlgorithm methodAlgorithm,
+                                              ReadyToRunFileLayoutAlgorithm fileAlgorithm,
+                                              ProfileDataManager profileData,
+                                              NodeFactory nodeFactory)
+        {
+            _methodLayoutAlgorithm = methodAlgorithm;
+            _fileLayoutAlgorithm = fileAlgorithm;
+            _profileData = profileData;
+            _nodeFactory = nodeFactory;
+        }
+
+        private ReadyToRunMethodLayoutAlgorithm _methodLayoutAlgorithm = ReadyToRunMethodLayoutAlgorithm.DefaultSort;
+        private ReadyToRunFileLayoutAlgorithm _fileLayoutAlgorithm = ReadyToRunFileLayoutAlgorithm.DefaultSort;
+        private ProfileDataManager _profileData;
+        private NodeFactory _nodeFactory;
+
+        private List<MethodWithGCInfo> SortByRegion(ImmutableArray<DependencyNodeCore<NodeFactory>> nodes, Func<MethodWithGCInfo, int> getRegion)
+        {
+            List<MethodWithGCInfo> methods = new List<MethodWithGCInfo>();
+            foreach (var node in nodes)
+            {
+                if (node is MethodWithGCInfo method)
+                {
+                    methods.Add(method);
+                }
+            }
+            methods.MergeSortAllowDuplicates((MethodWithGCInfo left, MethodWithGCInfo right) => getRegion(left).CompareTo(getRegion(right)));
+            return methods;
+        }
+
+        public ImmutableArray<DependencyNodeCore<NodeFactory>> ApplyProfilerGuidedMethodSort(ImmutableArray<DependencyNodeCore<NodeFactory>> nodes)
+        {
+            if (_methodLayoutAlgorithm == ReadyToRunMethodLayoutAlgorithm.DefaultSort)
+                return nodes;
+
+            Func<List<MethodWithGCInfo>> sortedMethods = null;
```

Comment:
Non-blocking nit - considering we use sortedMethods only once after the if statement to populate a list, is there any reason why using these auxiliary lambdas is more elegant than directly defining a List<MethodWithGCInfo> at this point and filling it with elements in the individual conditional blocks?